### ISSUE TITLE
Terminal updates after monitor disable & enable cycle broken with #1963

### ISSUE THREAD

**Issue by d-e-s-o** [rust-windowing/winit#2123]
Hi,

Over in https://github.com/alacritty/alacritty/issues/5703 we have been trying to get to the root of a bug that manifests in the alacritty terminal not being redrawn properly after the monitor got disabled and enabled. Since this is a regression, I bisected it to 
125ee0b446eab48dda56390c96b8f1973d45b3e6 (https://github.com/rust-windowing/winit/pull/1963).

Is this a known issue?

**Comment by maroider** [comment]
Thanks for reporting this. As far as I'm aware, this hasn't been reported before.

**Comment by d-e-s-o** [comment]
I played around with the code some. If I make the following change:
```patch
--- src/platform_impl/linux/x11/event_processor.rs
+++ src/platform_impl/linux/x11/event_processor.rs
@@ -1182,7 +1182,7 @@ impl<T: 'static> EventProcessor<T> {
                                                 // If there all monitors are closed before, scale
                                                 // factor would be already changed to 1.0.
                                                 maybe_prev_scale_factor.unwrap_or(1.0),
-                                                new_monitor.scale_factor,
+                                                1.0,
                                                 width,
                                                 height,
                                                 &*window.shared_state.lock(),
```

everything works fine. So I printed out `new_monitor.scale_factor` and it's `1.4166666666666667` (also see the [log](https://github.com/alacritty/alacritty/files/7812481/Alacritty-12601.log) I attached to the aforementioned `alacritty` task). So, if I change the code as follows:
```patch
--- src/platform_impl/linux/x11/event_processor.rs
+++ src/platform_impl/linux/x11/event_processor.rs
@@ -1181,7 +1181,7 @@ impl<T: 'static> EventProcessor<T> {
                                             let (new_width, new_height) = window.adjust_for_dpi(
                                                 // If there all monitors are closed before, scale
                                                 // factor would be already changed to 1.0.
-                                                maybe_prev_scale_factor.unwrap_or(1.0),
+                                                maybe_prev_scale_factor.unwrap_or(1.4166666666666667),
                                                 new_monitor.scale_factor,
                                                 width,
                                                 height,
```
everything works. So it appears that because we are defaulting to a wrong scaling factor of `1.0` we screw up the window update, setting sizes that don't make sense.

**Comment by d-e-s-o** [comment]
Given that we evidently picked the wrong magic number, may I suggest we just pick whatever the new monitor has as a scaling factor? After all, it's not clear to me in what basis we would reasonably assume that it changed.
```patch
--- src/platform_impl/linux/x11/event_processor.rs
+++ src/platform_impl/linux/x11/event_processor.rs
@@ -1181,7 +1181,7 @@ impl<T: 'static> EventProcessor<T> {
                                             let (new_width, new_height) = window.adjust_for_dpi(
                                                 // If there all monitors are closed before, scale
                                                 // factor would be already changed to 1.0.
-                                                maybe_prev_scale_factor.unwrap_or(1.0),
+                                                maybe_prev_scale_factor.unwrap_or(new_monitor.scale_factor),
                                                 new_monitor.scale_factor,
                                                 width,
                                                 height,
```

**Comment by maroider** [comment]
`1.4166666666666667` seems like a very unfortunate scale factor. Are you certain that this is a reasonable scale factor for your monitor configuration?

**Comment by chrisduerr** [comment]
> 1.4166666666666667 seems like a very unfortunate scale factor. Are you certain that this is a reasonable scale factor for your monitor configuration?

This isn't uncommon at all when you calculate scale factor automatically based on dimensions. My system has `1.1666666666666667` for example, the `666...` is extremely common and normal due to how it is calculated (I'd assume that's what threw you off).

**Comment by d-e-s-o** [comment]
> the `666...` is due to how it is calculated (I'd assume that's what threw you off).

It's actually the distance from `1.0` (the default) that matters. If no other monitor of the same name is available, `winit` will assume `1.0` as the "old" scale factor.  Then, in `Window::adjust_for_dpi` it uses this old scale factor to calculate the new window bounds (which it embeds in the DPR event) as follows:
```
        let scale_factor = new_scale_factor / old_scale_factor;
        // ...

        let new_width = (width as f64 * scale_factor).round() as u32;
        let new_height = (height as f64 * scale_factor).round() as u32;

        (new_width, new_height)
```

With a "new" scale factor of `1.46667` we end up with `scale_factor` being `1.4666667 / 1.0 = 1.4666667`. So now we multiply the actual bounds by `1.466667` and get bogus stuff. Any "new" scale factor other than `1.0` will result in incorrect results -- to varying degrees of severity.

I don't think guessing an old scale factor can work. The only sensible solution from my perspective is to use the new bounds directly, which is what my proposal above does (though in a roundabout way that surely could be made more explicit).
